packages/aws-cdk/lib/api/bootstrap/bootstrap-environment.ts

```diff
@@ -1,70 +1,123 @@
+    }
+
+    return deployBootstrapStack(
+      await this.loadTemplate(params),
+      {},
+      environment,
+      sdkProvider,
+      options);
   }
-  if (params.cloudFormationExecutionPolicies?.length) {
-    throw new Error('--cloudformation-execution-policies can only be passed for the new bootstrap experience.');
+
+  /**
+   * Deploy CI/CD-ready bootstrap stack from template
+   *
+   * @experimental
+   */
+  private async defaultBootstrap(
+    environment: cxapi.Environment,
+    sdkProvider: SdkProvider,
+    options: BootstrapEnvironmentOptions = {}): Promise<DeployStackResult> {
+
+    const params = options.parameters ?? {};
+
+    if (params.trustedAccounts?.length && !params.cloudFormationExecutionPolicies?.length) {
+      throw new Error('--cloudformation-execution-policies are required if --trust has been passed!');
+    }
+
+    const bootstrapTemplate = await this.loadTemplate();
+
+    return deployBootstrapStack(
+      bootstrapTemplate,
+      {
+        FileAssetsBucketName: params.bucketName,
+        FileAssetsBucketKmsKeyId: params.kmsKeyId,
+        TrustedAccounts: params.trustedAccounts?.join(','),
+        CloudFormationExecutionPolicies: params.cloudFormationExecutionPolicies?.join(','),
+        Qualifier: params.qualifier,
+        PublicAccessBlockConfiguration: params.publicAccessBlockConfiguration || params.publicAccessBlockConfiguration === undefined ? 'true' : 'false',
+      },
+      environment,
+      sdkProvider,
+      options);
   }
-  if (params.qualifier) {
-    throw new Error('--qualifier can only be passed for the new bootstrap experience.');
+
+  private async customBootstrap(
+    environment: cxapi.Environment,
+    sdkProvider: SdkProvider,
+    options: BootstrapEnvironmentOptions = {}): Promise<DeployStackResult> {
+
+    // Look at the template, decide whether it's most likely a legacy or modern bootstrap
+    // template, and use the right bootstrapper for that.
+    const version = bootstrapVersionFromTemplate(await this.loadTemplate());
+    if (version === 0) {
+      return this.legacyBootstrap(environment, sdkProvider, options);
+    } else {
+      return this.defaultBootstrap(environment, sdkProvider, options);
+    }
   }
 
-  return deployBootstrapStack(
-    legacyBootstrapTemplate(params),
-    {},
-    environment,
-    sdkProvider,
-    options);
-}
-
-/**
- * Deploy CI/CD-ready bootstrap stack from template
- *
- * @experimental
- */
-export async function bootstrapEnvironment2(
-  environment: cxapi.Environment,
-  sdkProvider: SdkProvider,
-  options: BootstrapEnvironmentOptions = {}): Promise<DeployStackResult> {
-
-  const params = options.parameters ?? {};
-
-  if (params.trustedAccounts?.length && !params.cloudFormationExecutionPolicies?.length) {
-    throw new Error('--cloudformation-execution-policies are required if --trust has been passed!');
+  private async loadTemplate(params: BootstrappingParameters = {}): Promise<any> {
```

Comment:
_Nit:_ Is it worth caching this result? In the custom template case, this will be called twice back-to-back. Not that bootstrapping is the most time-sensitive operation, but I have a knee-jerk response to repeated I/O to ask about caching.

Reply:
I feel ya, but in this case I think it's not worth it. It's a small-ish file that's guaranteed to be on disk. The OS buffers should have our back here.